### PR TITLE
[executor] support pipelined chunk execution (for state sync v2 and db-restore)

### PR DESCRIPTION
**Pull request by msmouse**
## Motivation
To enable true pipelined transaction replaying (for both the state sync v2 and db-restore), a queue of executed chunks need to be tracked, because for each chunk to execute (part of it, we need to know the committed ledger view and the latest ledger view). Instead of warping pending chunks to pseudo blocks, this gives it a clean implementation.
Demonstrated pipelining works by updating the db-restore tool.

performance:

Replaying the first 1M txns on the mainnet directly off Novi's backup in Azure, with this command:

```
cargo run --release -p backup-cli --bin replay-verify -- \
--concurrent-downloads 8 \
--target-db-dir ~/work/temp/mainnet-restore \
--metadata-cache-dir ~/mdc \
command-adapter \
 --config ~/work/diem_repos/partners/helm/validator/files/backup/azure.toml
```

getting:

```
2021-12-01T00:42:10.384026Z [tokio-runtime-worker] INFO storage/backup/backup-cli/src/backup_types/transaction/restore.rs:307 Transactions replayed. {"accumulative_tps":4823.795304341473,"version":999999}
```

compared to previously:

```
2021-12-01T00:36:04.239116Z [main] INFO diem-move/diem-vm/src/adapter_common.rs:193 Executing block, transaction count: 10000 {"first_version":990001,"name":"execution","txn_id":0}
2021-12-01T00:36:07.718483Z [main] INFO storage/backup/backup-cli/src/backup_types/transaction/restore.rs:333 transaction restore succeeded.
2021-12-01T00:36:07.718508Z [main] DEBUG storage/backup/backup-cli/src/backup_types/transaction/restore.rs:443 Accumulative TPS: 3123
```

The newly extracted algorithms will also be used in further BlockExecutor refactor.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/diem/diem/blob/main/CONTRIBUTING.md#pull-requests)?
Y

## Test Plan
unit tests

### CODE REVIEW

**Comment by libra-action** [comment]
**This PR may have modified JSON-RPC server code.**

Breaking changes policy:

1. Changing JSON-RPC method signatures or removing/modifying fields
   in the response in /V1 will not be accepted.
2. Adding fields is ok.
3. Adding a new JSON-RPC method is ok.

Please ensure you have also done the following to update the docs:

1. Update json-rpc/API-CHANGELOG.md.
2. Add/update type/method documentation under json-rpc/docs.

File: storage/storage-interface/src/lib.rs
Comment (JoshLind) [inline storage/storage-interface/src/lib.rs:102]: nit: `into_latest_tree_state`?

File: storage/backup/backup-cli/src/utils/stream/try_buffered_x.rs
Comment (JoshLind) [inline storage/backup/backup-cli/src/utils/stream/try_buffered_x.rs:4]: It kinda sucks that we have to manage these forks (including the ones already in the codebase). Perhaps we could consider pushing this upstream, or starting a conversation with the developers to provide this functionality.

**Review by JoshLind — COMMENTED** [review]
Looks great @msmouse! You'll probably need a second pair of eyes over the SMT stuff, but the rest of it looks solid. I've added a bunch of nits/comments/questions. 😄

File: execution/executor/src/chunk_executor.rs
Comment (msmouse) [reply execution/executor/src/chunk_executor.rs:33]: I'm deduplicating by extracting out what's in common -- "executing transactions", "processing vm output", "applying vm output to a speculative view", "committing". And I'm separating the two executors because they manage the pending transactions in two different structures: a block tree, and a chunk queue.

File: execution/executor/src/chunk_executor.rs
Comment (msmouse) [reply execution/executor/src/chunk_executor.rs:35]: Yes, 

>  every time it detects that the queue is not empty
and the current design is every time a chunk is finished execution and enqueued, the execution thread/task need to inform the commission thread/task to call commit_chunk() ONCE (it errors out if the queue is empty)

File: execution/executor-types/src/executed_chunk.rs
Comment (msmouse) [reply execution/executor-types/src/executed_chunk.rs:45]: The executed_chunk is not committed yet?

File: execution/executor/src/chunk_executor.rs
Comment (msmouse) [reply execution/executor/src/chunk_executor.rs:111]: When save_transactions() fails, technically we can retry.

File: execution/executor/src/chunk_executor.rs
Comment (msmouse) [reply execution/executor/src/chunk_executor.rs:229]: It can be that we gave up what's left in the queue? these two x_and_commit_chunk() interfaces will be in tests only anyway :P

File: execution/executor/src/components/apply_chunk_output.rs
Comment (msmouse) [reply execution/executor/src/components/apply_chunk_output.rs:81]: It allocates memory :( I used to be diligent to make them (event keys) static, but looks like other developers think allocating the Vec at runtime is preferred.. and I wasn't diligent enough to find out why. Probably a follow up.

File: execution/executor/src/metrics.rs
Comment (msmouse) [reply execution/executor/src/metrics.rs:76]: It's an accidental paste. Thank you for the thorough review!!!

File: storage/backup/backup-cli/src/utils/stream/try_buffered_x.rs
Comment (msmouse) [reply storage/backup/backup-cli/src/utils/stream/try_buffered_x.rs:4]: https://github.com/rust-lang/futures-rs/issues/2528

File: execution/executor/src/chunk_executor.rs
Comment (msmouse) [reply execution/executor/src/chunk_executor.rs:111]: Oh also the other thread will deem a version committed that's not yet committed.

File: execution/executor/src/chunk_executor.rs
Comment (JoshLind) [reply execution/executor/src/chunk_executor.rs:33]: Makes sense! I do wonder if we could ever move to a world where we could eliminate/minimize the block vs. chunk difference, but this is a good improvement.

File: execution/executor-types/src/executed_chunk.rs
Comment (JoshLind) [reply execution/executor-types/src/executed_chunk.rs:45]: Aah, typo: I was just suggesting following the inner name `to_commit`, e.g., `events_to_commit`? Just a nit though, so feel free to ignore.

File: execution/executor/src/chunk_executor.rs
Comment (JoshLind) [reply execution/executor/src/chunk_executor.rs:35]: I'm curious why you think this model is cleaner? In my mind, it's almost like a publish subscribe queue, right? We have a publisher (execute) and a subscriber (commit)? From the client perspective, it seems odd that we'd require the publisher to inform the subscriber out-of-band. Anyway, it's a small difference -- don't feel super strongly 😄

File: execution/executor/src/components/apply_chunk_output.rs
Comment (JoshLind) [reply execution/executor/src/components/apply_chunk_output.rs:81]: Aah, makes sense. Yeah, don't worry about it here.

File: execution/executor/src/chunk_executor.rs
Comment (JoshLind) [reply execution/executor/src/chunk_executor.rs:229]: I'd think if we wanted to give up on the queue, we'd call `reset()` ourselves? Likewise, if it's test-only we could use test-only complication But yeah, don't worry too much about it. It's nitty.

File: execution/executor/src/chunk_executor.rs
Comment (JoshLind) [reply execution/executor/src/chunk_executor.rs:111]: Aah, gotcha!

**Comment by JoshLind** [comment]
FYI @msmouse, all of my comments are nits/questions and not blockers, so once you've fixed the merge conflicts and decided how to respond, this is good to go!

File: execution/executor/src/chunk_executor.rs
Comment (msmouse) [reply execution/executor/src/chunk_executor.rs:33]: My thinking is the queue is kept for state sync and db-restore, and after we don't have the DEC ;-), we remove the block managing ability (consensus manages it, and keeps ExecutedChunk on its block tree)

File: execution/executor/src/chunk_executor.rs
Comment (msmouse) [reply execution/executor/src/chunk_executor.rs:35]: The state sync and the db-restore can choose to drive the pipeline differently. db-restore use async streams and Tokio tasks, doesn't need to explicitly wait for the last chunk commission, etc; and the state sync might have a commission thread spawned, and it might need to know explicitly if the commission queue is drained, etc. 

The ChunkExecutor implements the data structure (an ExecutedChunk queue), and identify the critical section for making sure it's thread safe (the mutex and mutex inside the SMT). On top of it we can have different ways to drive the parallelism at runtime. If both drive it the same way, we can extract that to a common part as well.

Ok to be honest I just like smaller structures (for me to build). I agree we can be pretty flexible and change at will any time.

File: execution/executor-types/src/executed_chunk.rs
Comment (msmouse) [reply execution/executor-types/src/executed_chunk.rs:45]: That makes sense!

File: execution/executor/src/chunk_executor.rs
Comment (msmouse) [reply execution/executor/src/chunk_executor.rs:229]: I'd just recreate the executor in case it's given up.

Yeah, it's now used by state sync v1. Once you move away from it, the compiler will complain it's not being used, I think?

File: execution/executor/src/components/apply_chunk_output.rs
Comment (zekun000) [inline execution/executor/src/components/apply_chunk_output.rs:108]: why do we keep `to_discard`? is it used later?

File: execution/executor/src/components/apply_chunk_output.rs
Comment (zekun000) [inline execution/executor/src/components/apply_chunk_output.rs:171]: this looks weird

File: execution/executor/src/components/apply_chunk_output.rs
Comment (zekun000) [inline execution/executor/src/components/apply_chunk_output.rs:195]: account.hash()?

File: execution/executor/src/chunk_executor.rs
Comment (JoshLind) [reply execution/executor/src/chunk_executor.rs:35]: Makes sense, thanks @msmouse 😄

File: execution/executor/src/components/apply_chunk_output.rs
Comment (zekun000) [inline execution/executor/src/components/apply_chunk_output.rs:225]: the name seems confusing, you mean compute_txn_data?

**Review by JoshLind — APPROVED** [review]
Thanks @msmouse! LGTM for the state sync stuff!

File: execution/executor/src/components/chunk_output.rs
Comment (zekun000) [inline execution/executor/src/components/chunk_output.rs:37]: was there a timer for this before?

File: vm-validator/src/vm_validator.rs
Comment (zekun000) [inline vm-validator/src/vm_validator.rs:37]: remove the ouroboros dependencies as well?

File: storage/scratchpad/src/sparse_merkle/node.rs
Comment (zekun000) [inline storage/scratchpad/src/sparse_merkle/node.rs:88]: what is this for?

File: execution/executor-types/src/executed_chunk.rs
Comment (msmouse) [reply execution/executor-types/src/executed_chunk.rs:145]: Technically as long as there's no discarded and to-retry transactions "in between" it's a valid chunk. And otherwise when I combine a chain of chunks by repeatedly calling combine(), I'm checking most of them twice.

File: execution/executor/src/components/apply_chunk_output.rs
Comment (msmouse) [reply execution/executor/src/components/apply_chunk_output.rs:48]: "sort" as in the "sorting hat" in Harry Porter?

File: execution/executor/src/components/apply_chunk_output.rs
Comment (msmouse) [reply execution/executor/src/components/apply_chunk_output.rs:108]: Guess just to be used in `ensure_no_discarded()` for now..

File: execution/executor/src/components/apply_chunk_output.rs
Comment (msmouse) [reply execution/executor/src/components/apply_chunk_output.rs:171]: adding comment

File: execution/executor/src/components/apply_chunk_output.rs
Comment (msmouse) [reply execution/executor/src/components/apply_chunk_output.rs:195]: There's also CryptoHash being imported in this file.

File: execution/executor/src/components/apply_chunk_output.rs
Comment (msmouse) [reply execution/executor/src/components/apply_chunk_output.rs:225]: transaction_data is less expressive in my opinion.. (maybe call the type ExecutedTransaction instead of TransactionData?)

File: execution/executor/src/components/chunk_output.rs
Comment (msmouse) [reply execution/executor/src/components/chunk_output.rs:37]: no https://github.com/diem/diem/blob/2289c823fde8c9511e5d124e4e81ef1d9cf12e21/execution/executor/src/lib.rs#L531

will add the one from the BlockExecutor

Comment (msmouse) [reply execution/executor/src/components/chunk_output.rs:37]: actually it's called "block execution timer". will add a separate one on the call site

File: storage/scratchpad/src/sparse_merkle/node.rs
Comment (msmouse) [reply storage/scratchpad/src/sparse_merkle/node.rs:88]: Even if we hold a reference to the oldest ancestor, there still can be nodes that's before it BEING dropped from another thread (the root has been dropped, but the nodes under it is BEING dropped.)

File: execution/executor/src/components/apply_chunk_output.rs
Comment (zekun000) [reply execution/executor/src/components/apply_chunk_output.rs:225]: what's the `diff` referred to? txn to be committed?

File: execution/executor-types/src/executed_chunk.rs
Comment (zekun000) [reply execution/executor-types/src/executed_chunk.rs:145]: I guess my question is more about where do we ensure the last piece? ideally we should have a CheckedChunk type that's transitioned from the raw one

File: vm-validator/src/vm_validator.rs
Comment (msmouse) [reply vm-validator/src/vm_validator.rs:37]: oh, looks like someone as good as me has done that already

File: execution/executor/src/components/apply_chunk_output.rs
Comment (msmouse) [reply execution/executor/src/components/apply_chunk_output.rs:227]: all the content (txns, txn outputs, state roots) needed to advance the ledger to a newer version

**Review by zekun000 — DISMISSED** [review]

File: execution/executor-types/src/executed_chunk.rs
Comment (msmouse) [reply execution/executor-types/src/executed_chunk.rs:145]: It's (already double) checked in `<ChunkExecutor as TransactionReplayer>::replay()`, this here is just to be complete (and politically right).

The block executor doesn't require a no-retry no-discard chunk to be accepted, so a CheckedChunk is not universal.

Probably just let ExecutedChunk not carry the to_retry and to_discard. I'll add a new commit.

Comment (msmouse) [reply execution/executor-types/src/executed_chunk.rs]: how about now

**Review by zekun000 — APPROVED** [review]

**Pull request closed without merging**
